File: algorithms/0121.best-time-to-buy-and-sell-stock/solution.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  max_profit = 0
  len = prices.length
  for i in 0..(len-2)
    max = prices[i+1...len].max
    diff = max - prices[i]
    max_profit = diff if diff > max_profit && diff.positive?
  end
  max_profit
end

File: algorithms/0121.best-time-to-buy-and-sell-stock/solution.go
func maxProfit(prices []int) int {
	maxProfit := 0
	buyPrice := prices[0]

	for _, price := range prices {
		if price < buyPrice {
			buyPrice = price
		}

		if maxProfit < price-buyPrice {
			maxProfit = price - buyPrice
		}
	}

	return maxProfit
}
